JwtModule secret 설정 추가로 WebSocket 인증 오류 해결

yxxjxxeee opened the pull request:
## #️⃣ 연관된 이슈
- #36

## 📝 작업 내용
- ChatGateway WebSocket JWT 인증 버그 수정 (secret 및 signOptions 설정 추가)
- WebSocket 네임스페이스 연결 및 메시지 송수신 e2e 테스트 추가
- CORS origin 설정 조정

yxxjxxeee committed:
fix: (#36) JwtModule secret 설정 추가로 WebSocket 인증 오류 해결

- JwtModule.register()에 JWT_ACCESS_SECRET, expiresIn 설정 추가
- WebSocket 네임스페이스 연결 및 메시지 송수신 e2e 테스트 추가
- CORS 설정 조정

yxxjxxeee committed:
test: (#36) 테스트 코드 수정

kimjiwonpg98 commented on src/chat/chat.gateway.ts:
```diff
@@ -42,7 +45,7 @@ export class ChatGateway implements OnGatewayConnection, OnGatewayDisconnect {
   private getClientData(client: Socket): ClientData {
     if (!client.data) {
-      client.data = { userId: '', nickname: '' };
+      throw new WsException('Client not authenticated');
```
data가 없는건데 인증이 안된게 맞을까요?

yxxjxxeee replied:
인증에 성공해야만 client.data가 설정되도록 구현해두었기 때문에
client.data가 없다는 것은 인증이 완료되지 않은 상태라는 의미로 작성했습니다.

kimjiwonpg98 replied:
getClientData라는 함수에서는 인증에 대한 어떤 행위든 관심이 없어야될 것 같은데 
인증 로직에서 바로 에러를 뱉어주는건 어떤가요?

data가 없음에는 다른 에러 매핑이 좋아보이긴합니다.

yxxjxxeee replied:
조언 감사합니다!
말씀해주신 것처럼 getClientData는 단순히 클라이언트 데이터를 조회하는 역할만 가지고 있어야 하고 
인증 여부 판단은 handleConnection에서 모두 처리하는 것이 책임 분리 측면에서 더 명확한 것 같습니다.

rrgks6221 commented on test/app.e2e-spec.ts:
```diff
@@ -16,7 +16,11 @@ describe('AppController (e2e)', () => {
+
   it('/ (GET)', () => {
-    return request(app.getHttpServer()).get('/').expect(200).expect('Hello World!');
+    return request(app.getHttpServer()).get('/').expect(200).expect('Hello PollyCasso!');
```
e2e 테스트를 짜실거라면 이런경우에는 String type인지만 확인하거나 http status code 체크 정도만 하면 좋을거같아요.
message는 언제든 변경될 수 있는 값이고 return type이 String 인것과 status code가 200인지가 주요한 테스트 범위이기 떄문에
만약 유지보수 되지 않을 코드라면 파일 자체를 지워버리는것도 좋을 수 있어요

yxxjxxeee replied:
말씀하신 부분 이해했습니다.
E2E 테스트에서는 메시지 내용보다는 타입과 상태 코드만 확인하는 것이 유지보수 측면에서 안전하겠네요.
그런데 굳이 필요 없을 것 같아 해당 파일은 삭제하기로 했습니다.

NicoDora commented on src/chat/chat.gateway.ts:
```diff
@@ -72,7 +77,7 @@ export class ChatGateway implements OnGatewayConnection, OnGatewayDisconnect {
     this.getClientData(client);
```
만약 유저가 토큰이 없거나 유효하지 않은 상태로 소켓 연결을 시도하면 58줄 혹은 72줄에 의해 소켓 연결이 끊기고 `getClientData()`내부의  WsException이 발생할거 같은데 혹시 의도된 부분일까요?

yxxjxxeee replied:
제가 알고 있는 흐름으로는 토큰에 문제가 있는 경우 client.disconnect()가 호출되기 때문에
다음 단계의 이벤트 핸들러들이 실행되지 않아 getClientData()의 WsException이 발생하는 경우는 없을 것 같습니다.
혹시 제가 잘못 이해하고 있는 부분이 있다면 알려주시면 감사하겠습니다!

NicoDora replied:
제가 생각하기엔 
1. `handleConnection()`에서 인증에 실패함.
2. `handleDisconnect()`이 호출됨.
3.  `getClientData()`가 실행됨.
4. 소켓에 데이터가 있는지 확인. -> 인증에 실패한 사용자이므로 데이터가 존재하지 않음. 따라서 WsException이 발생함.

이렇게 흘러갈거 같은데, 혹시 만료된 토큰으로 테스트 했을때 WsException이 발생하는지 확인해주실 수 있나요?

yxxjxxeee replied:
만료된 토큰으로 테스트해봤는데 제가 확인한 결과는 WsException이 아니라 client.emit('error', { message: 'Invalid token' })만 호출되네요

NicoDora commented on test/chat.gateway.e2e-spec.ts:
토큰 없이 소켓 연결 시도하는 경우, 유효하지 않은 토큰으로 소켓 연결을 시도하는 경우를 추가해서 의도대로 클라이언트가 에러 메시지를 받고 연결이 끊기는지에 대한 테스트 케이스가 추가되면 좋을거 같습니다!

yxxjxxeee replied:
좋은 것 같습니다!

yxxjxxeee committed:
refactor: (#36) e2e 소켓 인증 로직 테스트 및 getClientData 책임 분리

- 토큰 관련 e2e 테스트 케이스 추가
- getClientData 함수에서 인증 관련 로직 제거, 단순히 클라이언트 데이터 조회만 담당
- handleConnection에서 인증 처리하도록 책임 분리

yxxjxxeee committed:
Merge branch 'develop' of https://github.com/modern-agile-team/10term-pollycasso-back into fix/#36/websocket-jwt-verification

yxxjxxeee committed:
refactor: (#36) WebSocket 토큰 처리 및 에러 응답 형식 개선

- auth.token 및 Authorization 헤더에서 모두 토큰을 추출하도록 수정
- 로그 메시지 추가
- 에러 응답 메시지 형식 통일 및 구조 정리

yxxjxxeee committed:
Merge branch 'develop' of https://github.com/modern-agile-team/10term-pollycasso-back into fix/#36/websocket-jwt-verification

yxxjxxeee committed:
fix: (#36) 충돌 해결 및 테스트 코드 삭제
